src/main/java/com/agnieszkapawska/flashcards/domain/facades/LearningFacade.java

```diff
@@ -0,0 +1,122 @@
+                    flashcardService.saveFlashcard(flashcard);
+                });
+        if (shouldReturnFromMethod) {
+            return;
+        }
+
+        flashcardsToRepeatService.findByFlashcardIs(flashcardId)
+                .ifPresent(flashcardsToRepeat -> {
+                    markAnswer(answer, flashcard, flashcardsToRepeat, flashcardsToRepeatService, flashcardsToRefreshService);
+                    flashcardService.saveFlashcard(flashcard);
+                });
+        if (shouldReturnFromMethod) {
+            return;
+        }
+
+        flashcardsToRefreshService.findByFlashcardId(flashcardId)
+                .ifPresent(flashcardsToRefresh -> {
+                    markAnswer(answer, flashcard, flashcardsToRefresh, flashcardsToRefreshService, flashcardsToRefreshService);
+                    flashcardService.saveFlashcard(flashcard);
+                });
+    }
+
+    private void markAnswer(
+            Answer answer,
+            Flashcard flashcard,
+            FlashcardsStorage flashcardsStorage,
+            FlashcardsStorageService currentFlashcardsStorageService,
+            FlashcardsStorageService futureFlashcardsStorageService) {
+
+        if(answer.getIsCorrect()) {
+            if (flashcard.getCorrectAnswerCounter() < 2) {
+                flashcard.setCorrectAnswerCounter(flashcard.getCorrectAnswerCounter() + 1);
+            } else {
+                if(flashcardsStorage instanceof FlashcardsToRefresh) {
+                    shouldReturnFromMethod = true;
+                    return;
+                }
+                moveFlashcardFromOneFlashcardsStorageToAnother(
+                        answer, flashcard, flashcardsStorage, currentFlashcardsStorageService, futureFlashcardsStorageService);
+            }
+        } else {
+            flashcard.setCorrectAnswerCounter(0);
+            if (flashcardsStorage instanceof FlashcardsToLearn) {
+                shouldReturnFromMethod = true;
+                return;
+            }
+            moveFlashcardFromOneFlashcardsStorageToAnother(
+                    answer, flashcard, flashcardsStorage, currentFlashcardsStorageService, flashcardsToLearnService);
+        }
+        shouldReturnFromMethod = true;
+    }
+
+    private void moveFlashcardFromOneFlashcardsStorageToAnother(
+            Answer answer, Flashcard flashcard, FlashcardsStorage fromWhichFlashcardsStorageIsMoved,
+            FlashcardsStorageService currentFlashcardsStorageService, FlashcardsStorageService futureFlashcardsStorageService) {
+        Long userId = Long.parseLong(answer.getUserId());
+
+        fromWhichFlashcardsStorageIsMoved.getFlashcards().remove(flashcard);
+        currentFlashcardsStorageService.save(fromWhichFlashcardsStorageIsMoved);
+
+        FlashcardsStorage toWhichFlashcardsStorageIsMoved = (FlashcardsStorage) futureFlashcardsStorageService.findByUserId(userId)
+                .orElse(
```

Comment:
orElse vs orElseGet:
- orElse - to this variant we are passing object. Code is always called. Even if optional is empty.
- orElseGet - to this variant we are passing lambda. Code is called only if optional is empty.
So in your current implementation you are always creating new object for user. Even if he already have such storage already created.